### PR TITLE
Groupaccess support (https://github.com/PowerShell/Win32-OpenSSH/issues/463)

### PR DESCRIPTION
**Pull request by bingbing8**
https://github.com/PowerShell/Win32-OpenSSH/issues/463

### CODE REVIEW

File: contrib/win32/openssh/appveyor.psm1 [hunk 1]
```diff
@@ -494,43 +500,15 @@ function Deploy-OpenSSHTests
...
-    [System.IO.DirectoryInfo] $repositoryRoot = Get-RepositoryRoot    
+    [System.IO.DirectoryInfo] $repositoryRoot = Get-RepositoryRoot
     
     $sourceDir = Join-Path $repositoryRoot.FullName -ChildPath "regress\pesterTests"
-    Copy-Item -Path "$sourceDir\*" -Destination $OpenSSHTestDir -Include *.ps1,*.psm1 -Force -ErrorAction Stop
+    Copy-Item -Path "$sourceDir\*" -Destination $OpenSSHTestDir -Include *.ps1,*.psm1, sshd_config -Force -ErrorAction Stop
 
-    $sourceDir = Join-Path $repositoryRoot.FullName -ChildPath "bin\$folderName\$RealConfiguration"    
+    $sourceDir = Join-Path $repositoryRoot.FullName -ChildPath "bin\$folderName\$RealConfiguration"
     Copy-Item -Path "$sourceDir\*" -Destination $OpenSSHTestDir -Exclude ssh-agent.exe, sshd.exe -Force -ErrorAction Stop
```
Comment (manojampalam) [inline contrib/win32/openssh/appveyor.psm1]: Why are only ssh-agent.exe and sshd.exe included here? Why not all product binaries?

File: contrib/win32/openssh/appveyor.psm1 [hunk 2]
```diff
@@ -494,43 +500,15 @@ function Deploy-OpenSSHTests
...
-    if($env:psPath)
-    {
-        $strNewsubSystem = @"
-Subsystem	sftp	sftp-server.exe
-Subsystem	powershell	$env:psPath
-"@
-    }
-
-    (Get-Content $sshdConfigFile).Replace($strToReplace, $strNewsubSystem) | Set-Content $sshdConfigFile
+    Restart-Service sshd
```
Comment (manojampalam) [inline contrib/win32/openssh/appveyor.psm1]: Why is this needed?

File: groupaccess.c [hunk 3]
```diff
@@ -49,6 +52,73 @@ static char **groups_byname;
...
+	DWORD i = 0, j = 0;
+	LPLOCALGROUP_USERS_INFO_0 local_groups_info = NULL, tmp_groups_info;
+	wchar_t *user_utf16 = NULL, *full_name_utf16 = NULL, *udom_utf16 = NULL, *tmp;
+	DWORD entries_read = 0, total_entries = 0, full_name_len = 0, index = 0;
+	NET_API_STATUS nStatus;
+	
+	if (ngroups > 0)
+		ga_free();
+
+	user_utf16 = utf8_to_utf16(user);
```
Comment (manojampalam) [inline groupaccess.c:66]: NULL check?

File: groupaccess.c [hunk 4]
```diff
@@ -49,6 +52,73 @@ static char **groups_byname;
...
+	if ((tmp = wcschr(user_utf16, L'@')) != NULL) {
+		udom_utf16 = tmp + 1;
+		*tmp = L'\0';
+		index = wcslen(udom_utf16) + 1;
+		wmemcpy(full_name_utf16, udom_utf16, index);
+		full_name_utf16[wcslen(udom_utf16)] = L'\\';
+	}
+	wmemcpy(full_name_utf16 + index, user_utf16, wcslen(user_utf16) + 1);
+
+	nStatus = NetUserGetLocalGroups(NULL,
```
Comment (manojampalam) [inline groupaccess.c:86]: Does this return domain groups too?

File: groupaccess.c [hunk 5]
```diff
@@ -49,6 +52,73 @@ static char **groups_byname;
...
+		&total_entries);
+
+	if (NERR_Success != nStatus) {
+		error("NetUserGetLocalGroups() failed with error: %u",
+			nStatus);
+		errno = ENOENT;
+		goto done;
+	}
+
+	if (entries_read != total_entries) {
```
Comment (manojampalam) [inline groupaccess.c:102]: When will these be different?

File: groupaccess.c [hunk 6]
```diff
@@ -49,6 +52,73 @@ static char **groups_byname;
...
+		    "total_entries (%u) for user %.100s", entries_read, total_entries, user);
+		errno = ENOENT;
+		goto done;
+	}
+
+	if ((tmp_groups_info = local_groups_info) != NULL) {
+		groups_byname = xcalloc(entries_read, sizeof(*groups_byname));
+		for (i = 0, j = 0; i < total_entries; i++)
+		{
+			groups_byname[j++] = utf16_to_utf8(tmp_groups_info->lgrui0_name);
```
Comment (manojampalam) [inline groupaccess.c]: Add NULL check

**Review by manojampalam — CHANGES_REQUESTED** [review]

File: contrib/win32/openssh/appveyor.psm1 (hunk 1 above)
Comment (bingbing8) [inline contrib/win32/openssh/appveyor.psm1]: This is Deploy-OpenSSHTests. the purpose is to deploy unit tests. but * wildcard includes all files. ssh-agent.exe, sshd.exe are excluded because the service is already started and there will be error if deploy again. I will change the pattern to only deploy unit tests to avoid confusion.

File: groupaccess.c (hunk 4 above)
Comment (bingbing8) [reply groupaccess.c:86]: No, only local groups are returned. but flag LG_INCLUDE_INDIRECT will guarantee the group still returned if the user belong to some other global group and the global group belong to the local group.

File: groupaccess.c (hunk 5 above)
Comment (bingbing8) [reply groupaccess.c:102]: I don't see the cases when differences (when not all available entries were enumerated). My guess is when the 6th parameter (buffer size) is not enough, not all available entries got enumerated. since I pass MAX_PREFERRED_LENGTH to this parameter indicate the function allocates the amount of memory required for the data. but still check to sure.

File: groupaccess.c (hunk 6 above)
Comment (bingbing8) [reply groupaccess.c]: sure

File: contrib/win32/openssh/appveyor.psm1 (hunk 2 above)
Comment (bingbing8) [reply contrib/win32/openssh/appveyor.psm1]: This is test deployment. Before test deployment, the product binaries already got installed and service got started.

File: contrib/win32/openssh/appveyor.psm1 (hunk 1 above)
Comment (bingbing8) [reply contrib/win32/openssh/appveyor.psm1]: This is Deploy-OpenSSHTests. the purpose is to deploy unit tests. but * wildcard includes all files (when develop this script, the name pattern of the unit test files are not unique).  ssh-agent.exe, sshd.exe are excluded because the service is already started and there will be error if deploy again. I will change the pattern ( to unittest-*) to only deploy unit tests to avoid confusion.

File: groupaccess.c (hunk 3 above)
Comment (bingbing8) [reply groupaccess.c:66]: fixed
